api/admin/controller.py

```diff
@@ -411,9 +416,61 @@ def complaints(self, identifier_type, identifier):
         
         return response
 
+    def roles(self):
+        """Return a mapping from MARC codes to contributor roles."""
+        # TODO: The admin interface only allows a subset of the roles
+        # listed in model.py since it uses the OPDS representation of
+        # the data, and some of the roles map to the same MARC code.
+        CODES = Contributor.MARC_ROLE_CODES
+        marc_to_role = dict()
+        for role in [
+            Contributor.ACTOR_ROLE,
+            Contributor.ADAPTER_ROLE,
+            Contributor.AFTERWORD_ROLE,
+            Contributor.ARTIST_ROLE,
+            Contributor.ASSOCIATED_ROLE,
+            Contributor.AUTHOR_ROLE,
+            Contributor.COMPILER_ROLE,
+            Contributor.COMPOSER_ROLE,
+            Contributor.CONTRIBUTOR_ROLE,
+            Contributor.COPYRIGHT_HOLDER_ROLE,
+            Contributor.DESIGNER_ROLE,
+            Contributor.DIRECTOR_ROLE,
+            Contributor.EDITOR_ROLE,
+            Contributor.ENGINEER_ROLE,
+            Contributor.FOREWORD_ROLE,
+            Contributor.ILLUSTRATOR_ROLE,
+            Contributor.INTRODUCTION_ROLE,
+            Contributor.LYRICIST_ROLE,
+            Contributor.MUSICIAN_ROLE,
+            Contributor.NARRATOR_ROLE,
+            Contributor.PERFORMER_ROLE,
+            Contributor.PHOTOGRAPHER_ROLE,
+            Contributor.PRODUCER_ROLE,
+            Contributor.TRANSCRIBER_ROLE,
+            Contributor.TRANSLATOR_ROLE,
+            ]:
+            marc_to_role[CODES[role]] = role
+        return marc_to_role
+
+    def languages(self):
+        """Return the supported language codes and their English names."""
+        return LanguageCodes.english_names
+
+    def media(self):
+        """Return the supported media types for a work and their schema.org values."""
+        return Edition.additional_type_to_medium
+
     def edit(self, identifier_type, identifier):
         """Edit a work's metadata."""
 
+        # TODO: It would be nice to use the metadata layer for this, but
+        # this code handles empty values differently than other metadata
+        # sources. When a staff member deletes a value, that indicates
+        # they think it should be empty. This needs to be indicated in the
+        # db so that it can overrule other data sources that set a value,
+        # unlike other sources which set empty fields to None.
+
```

Comment:
I thought about this for a while and the only solution I can see is to define a special string that can be put into the database and means "suppress any other value".

Reply:
I think the first special string doesn't have to be a string; it can be an object like `NoValue` in `PatronData`. That might make things clearer.

The second "special string" could also be an object that indicates a special out-of-band value (a string or whatever it appropriate) should be stored in the database.